File: Problem 4 - Largest Palindrome/Perl/palindrome.pl
#!/usr/bin/env perl
use Modern::Perl '2012';
use List::Util qw(max);

sub generate_largest_palindrome {
    my($args_ref) = @_;
    my $min = $args_ref->{'min'};
    my $max = $args_ref->{'max'};

    my @candidates;

    for (my $j = $max; $j >= $min; $j--) {
        for (my $k = $max; $k >= $min; $k--) {
            my $candidate = $j * $k;

            if ( "$candidate" eq reverse "$candidate" ) {
                push @candidates, $candidate;
                #say "Largest is the product of $j and $k: " . $candidate;
            }
        }
    }

    return max @candidates;
}

say generate_largest_palindrome({
    min => 100,
    max => 999,
});

=begin Explanation

=end
=cut

File: Problem 4 - Largest Palindrome/Python/palindrome.py
#!/usr/bin/env python

def generate_largest_palindrome(min_lim, max_lim):
    candidates = []

    # Using a variable named 'max' interferes with the built-in function max()!
    # http://stackoverflow.com/questions/5581719/python-min-apparently-not-accepting-list-of-ints
    for j in range(min_lim, max_lim, 1):
        for k in range(min_lim, max_lim, 1):
            candidate = j * k

            # extended slice operator; same as s.join(reversed(s))
            if str(candidate) == str(candidate)[::-1]:
                candidates.append(candidate)

    return max(candidates)

print generate_largest_palindrome(100, 999)
